### ISSUE TITLE
[DX] Document the usage of theme-settings.php

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#3391]
So, I worked on #3375 and specifically on http://drupal.org/node/2900373. The initial PR I filed was an almost verbatim copy-paste from the Drupal patch, and it included a new `core/modules/simpletest/tests/themes/test_theme/theme-settings.php` file:

```php
<?php
 /**
 * @file
 * Theme setting callbacks for the test_theme theme.
 */
 /**
 * Implements hook_form_FORM_ID_alter().
 */
function test_theme_form_system_theme_settings_alter(&$form, &$form_state) {
  $form['test_theme_checkbox'] = array(
    '#type' => 'checkbox',
    '#title' => 'Test theme checkbox',
    '#default_value' => theme_get_setting('test_theme_checkbox'),
  );
   // Force the form to be cached so we can test that this file is properly
  // loaded and the custom submit handler is properly called even on a cached
  // form build.
  $form_state['cache'] = TRUE;
  $form['#submit'][] = 'test_theme_form_system_theme_settings_submit';
}
 /**
 * Form submission handler for the test theme settings form.
 *
 * @see test_theme_form_system_theme_settings_alter()
 */
function test_theme_form_system_theme_settings_submit($form, &$form_state) {
  backdrop_set_message('The test theme setting was saved.');
}
```

...so I enabled the `test_theme` theme to test on my local, and although the form with the test checkbox was being rendered under the theme settings, I was getting a WSOD, and this error when submitting the form:

```
Error: Call to undefined function test_theme_form_system_theme_settings_submit() in form_execute_handlers() (line 1543 of /var/www/docroot/core/includes/form.inc).
```

Having not worked on themes much before, and specifically `theme-settings.php` files, I tried to figure out how other core and contrib Backdrop themes are doing this. So this is what worked in the end:

```php
<?php
 /**
 * @file
 * Theme setting file for the test_theme theme.
 */
 /**
 * Implements hook_form_FORM_ID_alter().
 */
$form['test_theme_checkbox'] = array(
  '#type' => 'checkbox',
  '#title' => 'Test theme checkbox',
  '#default_value' => theme_get_setting('test_theme_checkbox', 'test_theme'),
);
 // Force the form to be cached so we can test that this file is properly
// loaded and the custom submit handler is properly called even on a cached
// form build.
$form_state['cache'] = TRUE;
```

As part of the process of figuring out things, I also searched our API site for `form_system_theme_settings`, and I got the following:

<img width="428" alt="screen shot 2018-11-24 at 4 59 49 pm" src="https://user-images.githubusercontent.com/2423362/48965077-6afb6680-f00a-11e8-9d16-ca669a75489e.png">

...since the other 3 results seemed theme-specific, I went with [`hook_form_system_theme_settings_alter()`](https://api.backdropcms.org/api/backdrop/core%21modules%21system%21theme.api.php/function/hook_form_system_theme_settings_alter), but that did not seem that helpful, because the Drupal implementation seemed to include a `hook_form_system_theme_settings_alter()` function for providing the form, and another `hook_form_system_theme_settings_submit()` function (`HOOK_FORM_submit()` actually). I did some more searching, but the closest I could get was [`bartik_form_system_theme_settings_alter_submit()`](https://api.backdropcms.org/api/backdrop/core%21modules%21color%21color.legacy.inc/function/bartik_form_system_theme_settings_alter_submit), which only states that it has been deprecated since 1.3.0.

I also searched for `theme-settings.php`, which brought up only this Bartik-specific page: https://api.backdropcms.org/api/backdrop/core%21themes%21bartik%21theme-settings.php, which was also not very helpful.

Have I been searching for the wrong hooks?

Furthermore, there's noting mentioned about `theme-settings.php` in "[Converting themes to Backdrop from Drupal 7](https://api.backdropcms.org/converting-themes)", so I think that we should update the API documentation and make sure that when people search for `theme-settings.php` they are presented with a generic, non-theme-specific page that includes Drupal 7 vs. Backdrop code examples.

**Comment by herbdool** [comment]
I think it might have changed here https://github.com/backdrop/backdrop-issues/issues/130.

We should update this document too https://api.backdropcms.org/adding-theme-settings-advanced

**Comment by herbdool** [comment]
It looks to me that the alter function is *not* deprecated. It still gets called https://github.com/backdrop/backdrop/blob/1.x/core/modules/system/system.admin.inc though we don't need to use it to add theme settings.

```
// Process the theme and all its base themes. foreach ($theme_keys as $theme) { // Include the theme-settings.php file. $filename = BACKDROP_ROOT . '/' . str_replace("/$theme.info", '', $themes[$theme]->filename) . '/theme-settings.php'; if (file_exists($filename)) { require_once $filename; } // Call theme-specific settings. $function = $theme . '_form_system_theme_settings_alter'; if (function_exists($function)) { $function($form, $form_state); } }
```

**Comment by klonos** [comment]
You are right @herbdool. It seems that the respective bit of code is the same in D7: https://github.com/drupal/drupal/blob/7.x/modules/system/system.admin.inc#L572

So I guess that the documentation for the `theme-settings.php` files conversion should mention that `HOOK_form_system_theme_settings_alter()` gets automatically called for you. So when porting themes from D7, you might need to remove the hook implementation from `theme-settings.php` (but still leave the code inside the hoo in place). Right?

**Comment by herbdool** [comment]
@klonos right, though the code looks slightly different.

I came to a different conclusion. From the code it looks like you don't need to change it. You can leave the hook in the theme since it's getting called. But looks like using the hook is optional.
